EIPS/eip-7495.md

```diff
@@ -118,51 +122,113 @@ Merkleization `hash_tree_root(value)` of an object `value` is extended with:
+- For fields of type `Optional[T]` in the base type `B`, the corresponding fields in `MerkleizeAs[B]` may have type `T` / `MerkleizeAs[T]` (to make them required in the sub-type), or `Optional[T]` / `Optional[MerkleizeAs[T]]` (to keep them optional in the sub-type)
+- Fields of type `Optional[T]` in the base type `B` may be absent in `MerkleizeAs[B]` (to disallow their presence in the sub-type)
+
+Serialization of `MerkleizeAs[B]` is based on its own field order, disregarding the order of base type `B`. If base type `B` is a `StableContainer[N]`, the leading `Bitvector` is replaced by a sparse reprsentation that only includes information about optional fields, and also follows the field order in `MerkleizeAs[B]` instead of the order within the base type `B`. Bits for required fields of `MerkleizeAs[B]` as well as the zero-padding to capacity `N` are not included. If there are no optional fields in `MerkleizeAs[B]`, no `Bitvector` is omitted.
+
+Merkleization of `MerkleizeAs[B]` always follows the merkleization of base type `B`.
+
+`MerkleizeAs[B]` is considered ["variable-size"](https://github.com/ethereum/consensus-specs/blob/4afe39822c9ad9747e0f5635cca117c18441ec1b/ssz/simple-serialize.md#variable-size-and-fixed-size) iff it contains any `Optional[T]` or any "variable-size" fields.
 
 ```python
-# Defines the common merkleization format and a portable serialization format across variants
+# Defines the common merkleization format and a portable serialization format
 class Shape(StableContainer[4]):
     side: Optional[uint16]
     color: uint8
     radius: Optional[uint16]
 
 # Inherits merkleization format from `Shape`, but is serialized more compactly
-class Square(Variant[Shape]):
+class Square(MerkleizeAs[Shape]):
     side: uint16
     color: uint8
 
 # Inherits merkleization format from `Shape`, but is serialized more compactly
-class Circle(Variant[Shape]):
+class Circle(MerkleizeAs[Shape]):
     radius: uint16
     color: uint8
-```
 
-In addition, `OneOf[S]` is defined to provide a `select_variant` helper function for determining the `Variant[S]` to use when parsing `StableContainer` `S`. The `select_variant` helper function MAY incorporate environmental information, e.g., the fork schedule.
+# Defines a container with immutable scheme that contains two `StableContainer`
+class ShapePair(Container):
+    shape_1: Shape
+    shape_2: Shape
 
-```python
-class AnyShape(OneOf[Shape]):
-    @classmethod
-    def select_variant(cls, value: Shape, circle_allowed = True) -> Type[Shape]:
-        if value.radius is not None:
-            assert circle_allowed
-            return Circle
-        if value.side is not None:
-            return Square
-        assert False
+# Inherits merkleization format from `ShapePair`, and serializes more compactly
+class SquarePair(MerkleizeAs[ShapePair]):
+    shape_1: Square
+    shape_2: Square
+
+# Inherits merkleization format from `ShapePair`, and reorders fields
```

Comment:
IMO we should consider removing MerkleizeAs["a container"] if the only benefit here is to support field reordering. (and 

In this example, CirclePair can just be a normal container like so and keep the same merkleization. If merkleization matching needs to be enforced, it can be done by external tooling or manual verification.
```python
class CirclePair(Container):
    shape_1: Circle
    shape_2: Circle
```

Also I'm just questioning if reordering of fields is helpful in general, or if its just asking too much from ssz. Is there a case where reordering of fields during serialization is required, and not just a nice-to-have?

Removing the bitvector for serialization by creating a "subtype" is helpful and has tangible benefit. It follows logically from motivation to StableContainer definition. It makes MerkleizeAs serialization more efficient and on-par with Containers. But reordering of fields feels more handwavey and wrt spec, looks more like a feature 'because we can'. My recommendation would be an additional EIP that extends functionality of MerkleizeAs, or defines a new thing with that functionality -- where the functionality is matched by the motivatoin.